File: frontend/src/types/mini-app.ts
export interface TelegramUser {
  id: number
  telegram_user_id: number
  first_name?: string
  last_name?: string
  username?: string
  language_code?: string
  is_premium?: boolean
  is_bot: boolean
  account_creation_date?: string
  created_at: string
  updated_at: string
}

export interface TelegramUserVerifyRequest {
  init_data: string  // Telegram Web App initData for verification
}

export interface TelegramUserVerifyResponse {
  verified: boolean
  telegram_user_id: number
  message: string
  user_data?: {
    id: number
    username?: string
    first_name?: string
    last_name?: string
  }
}

export interface UserSearchRequest {
  query: string
  limit?: number
  offset?: number
}

export interface UserHistoryEntry {
  id: number
  field_name: string
  old_value?: string | null
  new_value?: string | null
  changed_at: string
}

export interface UserSearchResult {
  telegram_user_id: number | string  // Can be number or masked string
  real_telegram_user_id?: number | null  // Real ID for masked accounts
  username?: string | null
  first_name?: string | null
  last_name?: string | null
  language_code?: string | null
  is_premium: boolean
  is_bot: boolean
  account_creation_date?: string | null
  created_at: string
  updated_at: string
  history: UserHistoryEntry[]
}

export interface UserSearchResponse {
  results: UserSearchResult[]
  total: number
  limit: number
  offset: number
}

// Telegram Web App types
declare global {
  interface Window {
    Telegram: {
      WebApp: {
        initData: string
        initDataUnsafe: {
          user?: {
            id: number
            first_name?: string
            last_name?: string
            username?: string
            language_code?: string
            is_premium?: boolean
            is_bot?: boolean
          }
          chat?: any
          query_id?: string
        }
        ready(): void
        close(): void
        expand(): void
        MainButton: {
          text: string
          show(): void
          hide(): void
          onClick(callback: () => void): void
          offClick(callback: () => void): void
        }
        BackButton: {
          show(): void
          hide(): void
          onClick(callback: () => void): void
          offClick(callback: () => void): void
        }
        HapticFeedback: {
          impactOccurred(style: 'light' | 'medium' | 'heavy' | 'rigid' | 'soft'): void
          notificationOccurred(type: 'error' | 'success' | 'warning'): void
          selectionChanged(): void
        }
        colorScheme: 'light' | 'dark'
        themeParams: {
          bg_color?: string
          text_color?: string
          hint_color?: string
          link_color?: string
          button_color?: string
          button_text_color?: string
        }
      }
    }
  }
}

File: backend/app/schemas/mini_app.py
"""
Mini app Pydantic schemas
"""

from pydantic import BaseModel
from typing import Optional, List, Union
from datetime import datetime


class TelegramUserVerifyRequest(BaseModel):
    """Schema for Telegram user verification from mini app"""
    init_data: str  # Telegram Web App initData for verification


class TelegramUserVerifyResponse(BaseModel):
    """Schema for Telegram user verification response"""
    verified: bool
    telegram_user_id: int
    message: str
    user_data: Optional[dict] = None
    session_token: Optional[str] = None  # Session token for frontend


class UserSearchRequest(BaseModel):
    """Schema for user search request"""
    query: str
    limit: Optional[int] = 20
    offset: Optional[int] = 0
    telegram_user_id: int  # User performing the search


class UserHistoryEntry(BaseModel):
    """Schema for user history entry"""
    id: int
    field_name: str
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    changed_at: datetime


class UserSearchResult(BaseModel):
    """Schema for individual user search result"""
    telegram_user_id: Union[int, str]  # Can be int or masked string
    real_telegram_user_id: Optional[int] = None  # Real ID for masked accounts
    username: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    language_code: Optional[str] = None
    is_premium: bool = False
    is_bot: bool = False
    account_creation_date: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    history: List[UserHistoryEntry] = []


class UserSearchResponse(BaseModel):
    """Schema for user search response"""
    results: List[UserSearchResult]
    total: int
    limit: int
    offset: int


class SearchLimitResponse(BaseModel):
    """Schema for search limit information"""
    total_searches_today: int
    max_searches_per_day: int
    remaining_searches: int
    reset_time: datetime  # When the limit will reset
    boost_searches_available: int = 0  # Additional searches available from purchases
    can_purchase_boost: bool = True  # Whether user can buy more boosts today


class SearchStatsEntry(BaseModel):
    """Schema for individual search statistics entry"""
    user_id: int
    telegram_user_id: int
    username: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    total_searches: int
    last_search_at: datetime
    searches_today: int


class SearchStatsResponse(BaseModel):
    """Schema for admin search statistics response"""
    total_users: int
    total_searches_all_time: int
    total_searches_today: int
    stats: List[SearchStatsEntry]


# Search Boost (Additional Searches) Schemas

class SearchBoostPriceResponse(BaseModel):
    """Schema for search boost price information"""
    id: int
    boost_amount: int  # Number of additional searches (default: 10)
    price_stars: int  # Price in Telegram Stars
    currency: str = 'XTR'
    is_active: bool
    created_at: datetime
    updated_at: datetime


class SearchBoostPriceUpdate(BaseModel):
    """Schema for updating search boost price"""
    price_stars: int


class SearchBoostAvailabilityResponse(BaseModel):
    """Schema for checking if user can purchase more boosts"""
    can_purchase: bool
    purchases_today: int
    max_purchases_per_day: int = 2
    remaining_purchases: int
    reason: Optional[str] = None


class SearchBoostPurchaseRequest(BaseModel):
    """Schema for initiating search boost purchase"""
    telegram_user_id: int


class SearchBoostPurchaseResponse(BaseModel):
    """Schema for search boost purchase response"""
    id: int
    user_id: int
    telegram_user_id: int
    boost_amount: int
    price_stars: int
    purchased_at: datetime
    used_searches: int
    is_active: bool
